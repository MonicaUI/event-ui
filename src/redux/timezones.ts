import reducer, { addTimezone, utcToDisplayDate } from 'event-widgets/redux/modules/timezones';
import { DEVICE_TIMEZONE_ID } from 'event-widgets/redux/selectors/timezone';
import { registerTranslation } from 'nucleus-guestside-site/src/redux/modules/text';

import { isPostRegistrationPage } from './website/registrationProcesses';
import { getCurrentPageId } from './pathInfo';
import { getRegCart } from './selectors/shared';
import { getPrimaryRegistrationId, getAttendee } from './registrationForm/regCart/selectors';
import { setSelectedTimeZone } from './timeZoneSelection';
import { convertEventTimezoneTranslations } from 'event-widgets/redux/selectors/timezone';
import { UPDATE_REG_CART_SUCCESS } from './registrationForm/regCart/actionTypes';

const DEVICE_TIMEZONE = 1;
const EVENT_TIMEZONE = 0;

/*
 * Creates a thunked action to load the timezone. Any language resource strings will
 * added to the necessary text resolvers.
 * @param The id of the timezone to load.
 */
export const loadTimezone =
  (timezoneId: $TSFixMe) =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      clients: { lookupClient },
      text
    } = getState();
    const locales = [text.locale];
    const response = await lookupClient.getTimezone(timezoneId, locales);
    const { timeZones, translations } = response;
    dispatch(addTimezone(timezoneId, timeZones[timezoneId]));
    dispatch(registerTranslation(translations[text.locale], text.locale, text.resolver.context));
  };

export const updateTimeZonePreference = (selectedTimeZone: $TSFixMe) => {
  return (dispatch: $TSFixMe, getState: $TSFixMe): $TSFixMe => {
    const timeZonePreference = selectedTimeZone.value === DEVICE_TIMEZONE_ID ? DEVICE_TIMEZONE : EVENT_TIMEZONE;
    const state = getState();
    const regCart = getRegCart(state);
    const {
      clients: { inviteeSearchClient }
    } = state;
    const isPostRegPage = isPostRegistrationPage(state, getCurrentPageId(state));
    const primaryEventRegistrationId = getPrimaryRegistrationId(regCart);
    const attendee = getAttendee(regCart, primaryEventRegistrationId);
    if (primaryEventRegistrationId && attendee) {
      const updatedRegCart = {
        ...regCart,
        eventRegistrations: {
          ...regCart.eventRegistrations,
          [primaryEventRegistrationId]: {
            ...regCart.eventRegistrations[primaryEventRegistrationId],
            attendee: {
              ...regCart.eventRegistrations[primaryEventRegistrationId].attendee,
              timeZonePreference
            }
          }
        }
      };
      dispatch({
        type: UPDATE_REG_CART_SUCCESS,
        payload: {
          regCart: updatedRegCart
        }
      });
    }

    if (isPostRegPage) {
      inviteeSearchClient.updateInviteeTimeZonePreference(attendee.attendeeId, timeZonePreference);
    }
  };
};

/**
 * Creates a thunked action to load the timezone data and register related translations for the event
 * into the redux store
 * @param timezoneId
 * @param timeZones
 * @param translations
 * @returns {Function}
 */
export const loadEventTimezoneIntoStore = (timezoneId: $TSFixMe, timeZones: $TSFixMe, translations: $TSFixMe) => {
  return (dispatch: $TSFixMe, getState: $TSFixMe): $TSFixMe => {
    const { text } = getState();
    dispatch(addTimezone(timezoneId, timeZones[timezoneId]));
    dispatch(registerTranslation(translations[text.locale], text.locale, text.resolver.context));
  };
};

/**
 * Update the selected Time Zone in state as per the
 * invitee time zone preference
 * @param {*} timezonePreference
 */
const updateSelectedTimeZone = timezonePreference => {
  return (dispatch, getState) => {
    const {
      event: { timezone },
      timezones,
      text: { translate }
    } = getState();
    const eventTimeZone = timezones[timezone];
    let preferedTimeZone;
    if (timezonePreference === DEVICE_TIMEZONE) {
      preferedTimeZone = {
        utcOffset: -new Date().getTimezoneOffset(),
        value: DEVICE_TIMEZONE_ID,
        name: translate('Event_GuestSide_DeviceTimeZone_desc_resx'),
        nameResourceKey: 'Event_GuestSide_DeviceTimeZone__resx'
      };
    } else {
      preferedTimeZone = {
        ...convertEventTimezoneTranslations(translate, eventTimeZone),
        value: timezone
      };
    }
    dispatch(setSelectedTimeZone(preferedTimeZone));
  };
};

export const updateSelectedTimeZoneForPrimaryAttendee = (regCart: $TSFixMe, currentEventRegistrationId: $TSFixMe) => {
  return (dispatch: $TSFixMe): $TSFixMe => {
    const primaryRegistrationAttendee = getAttendee(regCart, currentEventRegistrationId);
    // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
    if (primaryRegistrationAttendee && primaryRegistrationAttendee.timeZonePreference) {
      dispatch(updateSelectedTimeZone(primaryRegistrationAttendee.timeZonePreference));
    }
  };
};

/*
 * Loads the timezone for the event.
 */
export const loadEventTimezone =
  () =>
  async (dispatch: $TSFixMe, getState: $TSFixMe): Promise<$TSFixMe> => {
    const {
      event: { timezone }
    } = getState();
    await dispatch(loadTimezone(timezone));
  };

// Export the reducer and any methods from base timezone reducer for ease of use.
export default reducer;
export { utcToDisplayDate };
