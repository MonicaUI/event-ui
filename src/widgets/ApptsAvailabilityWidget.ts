import { connect } from 'react-redux';

import withMappedWidgetConfig from 'nucleus-widgets/lib/withMappedWidgetConfig';

import ApptsAvailabilityWidget from 'event-widgets/lib/ApptsAvailability/ApptsAvailabilityWidget';
import { getEventRegistrationId } from '../redux/selectors/currentRegistrant';
import { getPrimaryAndGuestSelectedSessions } from '../redux/selectors/productSelectors';
import { getAllSortedSessionsForWidget } from '../redux/selectors/productSelectors';
import { getRegCart } from '../redux/selectors/shared';
import { getIn } from 'icepick';
import { setAppointmentsAvailability } from '../redux/registrationForm/regCart/actions';

/* Returns appointments availability from reg cart */
export const getApptsAvailability = (state: $TSFixMe, eventRegistrationId: $TSFixMe): $TSFixMe => {
  const regCart = getRegCart(state);
  return getIn(regCart, ['eventRegistrations', eventRegistrationId, 'appointmentsAvailability']);
};

export default withMappedWidgetConfig(
  connect(
    (state: $TSFixMe, props: $TSFixMe) => {
      const selectedSessions = getPrimaryAndGuestSelectedSessions(state);
      const {
        event: { timezone },
        appointments: { appointmentEvent }
      } = state;
      const eventRegistrationId = getEventRegistrationId(state);
      // @ts-expect-error ts-migrate(2554) FIXME: Expected 1 arguments, but got 2.
      const sessionInfo = getAllSortedSessionsForWidget(state, 'ApptsAvailability');
      return {
        config: props.config,
        selectedSessions,
        sessionInfo,
        isGuest: true,
        timezone: state.timezones[timezone],
        apptsAvailability: getApptsAvailability(state, getEventRegistrationId(state)),
        eventRegistrationId,
        globalTheme: state.website.theme.global,
        appointmentEvent
      };
    },
    {
      onSetAppointmentsAvailability: setAppointmentsAvailability
    },
    (stateProps: $TSFixMe, dispatchProps: $TSFixMe, ownProps: $TSFixMe) => {
      return {
        ...ownProps,
        ...stateProps,
        onSetAppointmentsAvailability: dispatchProps.onSetAppointmentsAvailability
      };
    }
  )(ApptsAvailabilityWidget)
);
