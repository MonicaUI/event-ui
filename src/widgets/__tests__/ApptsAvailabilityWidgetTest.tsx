import React from 'react';
import ApptsAvailabilityWidget from '../ApptsAvailabilityWidget';
import { mount } from 'enzyme';
import { utcToDisplayDate } from 'event-widgets/redux/modules/timezones';

const timeZone = {
  id: 35,
  name: 'Eastern Time',
  nameResourceKey: 'Event_Timezone_Name_35__resx',
  plannerDisplayName: '(GMT-05:00) Eastern [US & Canada]',
  hasDst: true,
  utcOffset: -300,
  abbreviation: 'ET',
  abbreviationResourceKey: 'Event_Timezone_Abbr_35__resx',
  dstInfo: []
};

const subscribe = () => {};

const defaultProps = {
  classes: {},
  style: { palette: {} },
  translate: (res, opts) => (opts ? `${res}:${JSON.stringify(opts)}` : res),
  isRegistrationPage: true,
  id: 'widget:apptsAvailability',
  'data-cvent-id': 'widget-ApptsAvailability-widget:apptsAvailability',
  type: 'ApptsAvailability',
  store: { dispatch, getState, subscribe },
  config: {
    headerText: 'Appts Availability',
    instructionalText: 'Sample Instructions',
    appData: {
      availabilityConfigOption: 'unavailability',
      apptEventDayInfo: [
        {
          date: '2019-06-28',
          dateEnabled: false,
          startTime: '09:00',
          endTime: '17:00'
        },
        {
          date: '2019-06-29',
          dateEnabled: false,
          startTime: '09:00',
          endTime: '17:00'
        }
      ]
    }
  }
};

const loadMetaDataFunction = jest.fn(() => {
  return {};
});

function getState() {
  return {
    guestText: {
      translate: resx => resx,
      translateTime() {},
      translateDate() {},
      resolver: {}
    },
    appointments: {
      appointmentEvent: {
        isValid: true,
        startTime: utcToDisplayDate('2019-06-28T09:00:00.000Z', timeZone),
        endTime: utcToDisplayDate('2019-06-29T10:00:00.000Z', timeZone),
        timeZone
      }
    },
    timezones: {
      35: timeZone
    },
    event: {
      timezone: 35,
      products: {
        admissionItems: {
          admissionItemNoFee: {
            id: 'admissionItemNoFee',
            capacityId: 'admissionItem1',
            name: 'Admission Item No Fee',
            code: 'Admission Item Code 2',
            description: 'Admission Item Description',
            associatedOptionalSessions: [],
            defaultFeeId: '00000000-0000-0000-0000-000000000000',
            fees: {},
            displayOrder: 2,
            isOpenForRegistration: true,
            applicableContactTypes: []
          }
        },
        sessionContainer: {
          optionalSessions: {
            associatedSession: {
              capacityId: 'associatedSessionCapacityId'
            },
            fullAssociatedSession: {
              capacityId: 'fullAssociatedSessionCapacityId'
            }
          }
        }
      }
    },
    website: {
      theme: {
        global: {}
      },
      pluginData: {
        registrationProcessNavigation: {
          registrationPaths: {
            regProcessStep1: {
              pageIds: ['regProcessStep1', 'regProcessStep2']
            }
          }
        }
      },
      siteInfo: {
        sharedConfigs: {}
      },
      layoutItems: {}
    },
    widgetFactory: {
      loadMetaData: loadMetaDataFunction
    },
    appData: {
      registrationSettings: {
        registrationPaths: {
          defaultRegPath: {
            name: 'reg_path_1'
          },
          newRegPath: {
            name: 'reg_path_2'
          }
        }
      }
    },
    pathInfo: {
      currentPageId: 'regProcessStep2'
    },
    registrationForm: {
      currentEventRegistrationId: 'eventRegistrationId',
      regCart: {
        regCartId: 'regCartId',
        eventRegistrations: {
          eventRegistrationId: 'eventRegistrationId'
        }
      }
    },
    capacity: {},
    text: {
      resolver: {
        date: () => 'some date ',
        currency: x => x
      },
      translate: x => x
    },
    userSession: {},
    defaultUserSession: {
      isPlanner: false
    },
    config: {
      headerText: 'Appts Availability',
      instructionalText: 'Sample Instructions',
      availabilityConfigOption: 'unavailability',
      apptEventDayInfo: [
        {
          date: '2019-06-28',
          dateEnabled: false,
          startTime: '09:00',
          endTime: '17:00'
        },
        {
          date: '2019-06-29',
          dateEnabled: false,
          startTime: '09:00',
          endTime: '17:00'
        }
      ]
    },
    type: 'ApptsAvailability'
  };
}

async function dispatch(action) {
  if (typeof action === 'function') {
    await action(dispatch, getState);
  }
}

describe('ApptsAvailabilityWidget', () => {
  test('should match snapshot', () => {
    const widget = mount(<ApptsAvailabilityWidget {...defaultProps} />);
    expect(widget.props()).toMatchSnapshot();
  });
});
