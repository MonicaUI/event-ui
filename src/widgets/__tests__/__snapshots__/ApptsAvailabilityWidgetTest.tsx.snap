// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApptsAvailabilityWidget should match snapshot 1`] = `
Object {
  "classes": Object {},
  "config": Object {
    "appData": Object {
      "apptEventDayInfo": Array [
        Object {
          "date": "2019-06-28",
          "dateEnabled": false,
          "endTime": "17:00",
          "startTime": "09:00",
        },
        Object {
          "date": "2019-06-29",
          "dateEnabled": false,
          "endTime": "17:00",
          "startTime": "09:00",
        },
      ],
      "availabilityConfigOption": "unavailability",
    },
    "headerText": "Appts Availability",
    "instructionalText": "Sample Instructions",
  },
  "data-cvent-id": "widget-ApptsAvailability-widget:apptsAvailability",
  "id": "widget:apptsAvailability",
  "isRegistrationPage": true,
  "store": Object {
    "dispatch": [Function],
    "getState": [Function],
    "subscribe": [Function],
  },
  "style": Object {
    "palette": Object {},
  },
  "translate": [Function],
  "type": "ApptsAvailability",
}
`;
